src/lib/tools.ts

```diff
@@ -0,0 +1,314 @@
+ * @param {object} conf
+ * @param {string} basePath
+ * @returns {object}
+ */
+export function convertRelativePaths({
+	conf,
+	basePath,
+}: {
+	conf: any;
+	basePath: string;
+}): any {
+	// Convert relative to absolute URIs
+	const keys = [
+		'productRepo',
+		'archiveStore',
+		'encryptionKeyPath',
+		'envFile.path',
+		'yamlFile.path',
+		'kubernetes.kubeConfigPath',
+		'kubernetes.bastion.key',
+		'compose.socket',
+	];
+
+	for (const key of keys) {
+		const value = _.get(conf, key);
+		if (value) {
+			_.set(conf, key, getAbsoluteUri(value, basePath));
+		}
+	}
+	return conf;
+}
+
+/**
+ * Reads a file from URI
+ * @param {string} URI
+ * @param {string} path
+ * @param {string} cachePath
+ * @returns {Promise<string>}
+ */
+export async function readFromUri({
+	uri,
+	path,
+	cachePath,
+}: {
+	uri: string;
+	path: string;
+	cachePath?: string;
+}): Promise<string> {
+	// TODO: support git URI
+	if (isValidGitUri(uri)) {
+		throw new UnsupportedError('Git URI support not implemented yet');
+	} else if (localPathUri(uri)) {
+		return (await fs.readFile(join(uri, path))).toString('utf8');
+	} else {
+		throw new UnsupportedError('URI type not supported yet');
+	}
+}
+
+/**
+ * Lists a path in a URI
+ * @param {string} URI
+ * @param {string} path
+ * @param {string} cachePath
+ * @returns {Promise<string[]>}
+ */
+export async function listUri({
+	uri,
+	path,
+	cachePath,
+}: {
+	uri: string;
+	path: string;
+	cachePath?: string;
+}): Promise<string[]> {
+	// TODO: support git URI
+	if (isValidGitUri(uri)) {
+		throw new UnsupportedError('Git URI support not implemented yet');
+	} else if (localPathUri(uri)) {
+		return await fs.readdir(join(uri, path));
+	} else {
+		throw new UnsupportedError('URI type not supported yet');
+	}
+}
+
+/**
+ * Keyframe unwrapper to a standard format
+ * @param productRepoURI: URI of the product repo
+ * @param keyFramePath: path of keyframe
+ *
+ * returns: Keyframe object
+ */
+export async function unwrapKeyframe(
+	productRepoURI: string,
+	keyFramePath: string = './keyframe.yml',
+): Promise<object> {
+	// TODO: keyframe layering
+	let keyFrame = await loadFromUri({ uri: productRepoURI, path: keyFramePath });
+
+	if (keyFrame) {
+		keyFrame = _.filter(
+			_.get(keyFrame, ['children', 'sw', 'containerized-application'], []),
+			component => component.type === 'sw.containerized-application',
+		);
+		keyFrame = _.mapValues(_.keyBy(keyFrame, 'slug'), o =>
+			_.merge(_.get(o, 'assets', {}), { version: _.get(o, 'version') }),
+		);
+		return keyFrame;
+	} else {
+		return {};
+	}
+}
+
+/**
+ * Creates an ssh tunnel for executing a promise
+ * @param tnlConfig: ssh2 tunnel configuration object
+ * @param prom: promise
+ * @param timeout: tunnel timeout.
+ * @returns {Promise<any>}
+ */
+export async function runInTunnel(
+	tnlConfig: tunnel.Config,
+	prom: Promise<any>,
+	timeout: number,
+): Promise<any> {
+	const tnl = await tunnelAsync(tnlConfig);
```

Comment:
@Page- It is used by bringing up the SSH tunnel with a local port forward `8443` to the kubernetes host on the other side of the tunnel. The `prom` promise is run, whereby it uses the tunnel port `127.0.0.1:8443` to talk to the K8s host, then on completion it drops the tunnel.

Reply:
But the `prom` promise is already running, otherwise you wouldn't have a promise available yet, which means it starts before the tunnel is created.  So I thought it's unused because it starts too late but it sounds like it is instead is used but due to a race condition that allows it to be started before it's needed

Reply:
@Page- I have changed the signature so that it expects a *function* which returns a `Promise<any>` - so the promise will not be created until after the tunnel is up 👍

Reply:
I also side-tested this mechanism (passing a function to resolve to a promise) and it worked as expected - this should work in this case too